File: leetcode/java/SymmetricTree.java
import java.util.Stack;

/**
 * Created by Robbert on 1/23/15.
 * Symmetric Tree
 * https://oj.leetcode.com/problems/symmetric-tree/
 */
public class SymmetricTree {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) {
            return true;
        } else {
            return checkSymmetric(root.left, root.right);
        }
    }

    private boolean checkSymmetric(TreeNode left, TreeNode right) {
        if (left == null && right == null) {
            return true;
        }

        if (left == null || right == null || left.val != right.val) {
            return false;
        }

        return checkSymmetric(left.left, right.right)
                && checkSymmetric(left.right, right.left);
    }

    public boolean isSymmetricIteration(TreeNode root) {
        if (root == null) {
            return true;
        }

        Stack<TreeNode> stack = new Stack<TreeNode>();
        stack.push(root.right);
        stack.push(root.left);
        while (!stack.isEmpty()) {
            TreeNode left = stack.pop();
            TreeNode right = stack.pop();
            if (left == null && right == null) {
                continue;
            }
            if (left == null || right == null || left.val != right.val) {
                return false;
            }
            stack.push(right.left);
            stack.push(left.right);
            stack.push(right.right);
            stack.push(left.left);
        }

        return true;
    }
}

File: leetcode/python/symmetric_tree.py
#!/usr/bin/env python

# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isSymmetric(self, root):
        if not root:
            return True
        return self.symmetric(root.left, root.right)

    def symmetric(self, left, right):
        if not left and not right: # base case, hit the end
            return True

        if not left or not right: # subtrees have different depths
            return False
        if left.val != right.val: # subnodes' value aren't equal
            return False
        
        # recursively compare subtrees
        return self.symmetric(left.left, right.right) and \
            self.symmetric(left.right, right.left)

    def isSymmetric(self, root):
        '''Iteratly traverse the tree in preorder.'''
        if not root:
            return True
        stack = [root.left, root.right]
        while stack:
            rgt, lft = stack.pop(), stack.pop()
            if not lft and not rgt:
                continue
            if not lft or not rgt or lft.val != rgt.val:
                return False
            stack += [lft.left, rgt.right, lft.right, rgt.left]
        return True
